docs/madr/003-arns-undername-limits.md

```diff
@@ -0,0 +1,237 @@
+    if (sortOrder === 'desc') {
+      return a[sortKey] - b[sortKey];
+    } else {
+      return b[sortKey] - a[sortKey];
+    }
+  }
+  return a.localeCompare(b);
+});
+
+// enforce undername limit against sorted records, using the priority field, fallback to alphabetical
+```
+
+## Pros and Cons of Options
+
+### Option 1: ANTs provide sorted array of records
+
+- `+` Sort order is controlled entirely by ANTs
+- `+` Minimal changes to gateways (use new API)
+- `-` Additional changes in ar-io-sdk to support non updated ANTs
+- `-` Managing priority order when updating records (handling collisions)
+
+### Option 2: ANTs provide priority data in records
+
+- `+` Extends existing ANT records object type
+- `+` Minimal changes to [ar-io-node] & [ar-io-sdk] (easy to make backwards compatible since object response type is unchanged)
+- `+` Supports implementing Option #1 at a later time, if desired
+- `-` Managing priority order when updating records (handling collisions)
+- `-` Gateways must fetch full ANT records and sort them by priority order
+- `-` Gateways must compute the sorted array of records
+
+### Option 3: ANT provides sort attributes via separate API
+
+- `+` Same pros as Option #2
+- `-` Same cons as Option #2
+- `-` Additional api calls to get sort attributes
+
+## Decision
+
+We will implement Option #2: **Additional priority attribute in ANT records**
+
+Ultimately, we could choose to implement Option #1 and Option #2. Both require additional changes to the ANT contract, and the [ar-io-node]. Option #1 however, requires changes suggested in Option #2 AND additional modifications to return the new object type.
+
+Option #2 requires the least amount of changes, while satisfying the requirements for enforcing undername limits.
+
+Example ANT records state with priority data:
+
+```json
+{
+  "@": {
+    "transactionId": "0x123",
+    "ttlSeconds": 1000000,
+    "priority": 0
+  },
+  "undername1": {
+    "transactionId": "0x123",
+    "ttlSeconds": 1000000,
+    "priority": 1
+  },
+  "undername2": {
+    "transactionId": "0x123",
+    "ttlSeconds": 1000000,
+    "priority": 2
+  }
+}
+```
+
+The [ar-io-node] will sort the records by priority, and fallback to alphabetical sorting when the priority attribute is not present and enforce the undername limit against the sorted records. If ANTs provide invalid priority order (conflicting records with same priority), **the [ar-io-node] will return a 400 error to notify the ANT owner of invalid priority order.**
+
+Gateways will leverage existing caching logic defined in [./002-arns-cache-timing.md]. As such, the [ar-io-node] will cache the records (including priority order) for the provided TTLs. If the ANT contract is updated, the [ar-io-node] will fetch the updated records after the provided TTL and subsequently enforce any changes in priority order against the updated records.
+
+Next steps:
+
+- Update [ANT source code] to set priority order when updating/appending records
```

Comment:
the number does not represent a position, it only represents the number

Reply:
added note in [90be780](https://github.com/ar-io/ar-io-node/pull/272/commits/90be780969e021f3f9cfee168c331a143129ec1e)